c/tskit/trees.h

```diff
@@ -977,17 +976,141 @@ bool tsk_tree_is_descendant(const tsk_tree_t *self, tsk_id_t u, tsk_id_t v);
 @{
 */
 
-int tsk_tree_preorder(
-    const tsk_tree_t *self, tsk_id_t root, tsk_id_t *nodes, tsk_size_t *num_nodes);
-int tsk_tree_postorder(
+/**
+@brief Fill an array with the nodes of this tree in preorder.
+
+@rst
+Populate an array with the nodes in this tree in preorder. The array
+must be pre-allocated and be sufficiently large to hold the array
+of nodes visited. The recommended approach is to use the
+:c:func:`tsk_tree_get_size_bound` function, as in the following example:
+
+.. code-block:: c
+
+    static void
+    print_preorder(tsk_tree_t *tree)
+    {
+        int ret;
+        tsk_size_t num_nodes, j;
+        tsk_id_t *nodes = malloc(tsk_tree_get_size_bound(tree) * sizeof(*nodes));
+
+        if (nodes == NULL) {
+            errx(EXIT_FAILURE, "Out of memory");
+        }
+        ret = tsk_tree_preorder(tree, nodes, &num_nodes);
+        check_tsk_error(ret);
+        for (j = 0; j < num_nodes; j++) {
+            printf("Visit preorder %lld\n", (long long) nodes[j]);
+        }
+        free(nodes);
+    }
+
+.. seealso::
+    See the :ref:`sec_c_api_examples_tree_traversals` section for
+    more examples.
+
+@endrst
+
+@param self A pointer to a tsk_tree_t object.
+@param nodes The tsk_id_t array to store nodes in. See notes above for
+    details.
+@param num_nodes A pointer to a tsk_size_t value where we store the number
+    of nodes in the traversal.
+@return 0 on success or a negative value on failure.
+*/
+int tsk_tree_preorder(const tsk_tree_t *self, tsk_id_t *nodes, tsk_size_t *num_nodes);
+
+/**
+@brief Fill an array with the nodes of this tree starting from a particular node.
+
+@rst
+As for :c:func:`tsk_tree_preorder` but starting the traversal at a particular node.
+The value ``-1`` is treated as a special case, in which we visit all nodes reachable
+from the roots. The virtual root is also a valid input for this function.
```

Comment:
Maybe link to `sec-data-model-tree-roots` wherever we mention the virtual root?

Reply:
Good call, done. Ready to go now I think